### ISSUE TITLE
Stylelint: Consider defining Specific Properties Available

### ISSUE THREAD

**Issue by mikeselander** [humanmade/coding-standards#104]
We are currently whitelisting unit(less, actually)s for `line-height` only in our stylelint config, but I propose that we get more specific with what units are available.

### Font Size
Common best practice is to use `rem` units for font size based off of a single root value. We should enforce `rem` units, or at a minimum, blacklist `px` units to prevent them from being used.

### `ms` Units
For timing functions, such as transitions or animations, `ms` units are far clearer and easy to digest than seconds because they are laid out in longform. While they mean the same thing, `400ms` is cleaner than `.4s` and feels more precise in my opinion. I suggest that we enforce ms units on all timing functions.

Ref https://stylelint.io/user-guide/rules/declaration-property-unit-whitelist/ & https://stylelint.io/user-guide/rules/declaration-property-unit-blacklist/

Thoughts #teamfrontend?

cc @peterwilsoncc @sambulance @goldenapples @rmccue @joemcgill

**Comment by peterwilsoncc** [comment]
For font-size: I'd prefer to blacklist `px` rather than whitelist `rem`. `em` can be [useful within components](https://css-tricks.com/rem-global-em-local/) and other relative units such as `vw, vh` can be useful for responsive font sizes.

Timing functions: `ms` works for me.

I'm going to cc @kirstyburgoine as part of #teamfrontend

**Comment by rmccue** [comment]
I strongly agree that we should block `px` rather than only allow `rem`. The bane of my existence right now is the overuse of `rem` within the pattern library, which makes it much harder to use things like headers inside sidebars where the whole sidebar is `0.9rem`.

IMO, `rem` should be used for layout-level things, like sizing up/down sidebars, content areas, etc as well as page layout. `em` should be used for elements within blocks. A different way to think about this could be that `rem` should only be used with ID selectors, and `em` should be used elsewhere. (Although, I'd actually use classes rather than IDs typically...)

Agree on `ms`, and I haven't actually seen `s` much in the wild.

p.s. created a @humanmade/frontend team

**Comment by kirstyburgoine** [comment]
Thanks for cc'ing me @peterwilsoncc :) I definitely agree with blacklisting `px` as the approach. 

I think my approach to units is a bit different to @rmccue's. I use `rem` for all type sizing. This is because it takes its base value from the body font size making it an easier unit to work with. It is also needed when applying responsive type calculations to the body tag so that they are applied to all type sizing in your page. 

`em` is then used for all padding and spacing so that it can scale with the browser window, but not used for fonts. I find `em` can be very unpredictable for type sizing because it takes its base value from the parent container it is within instead of the body tag. If you had two containers in your layout with different font sizes set on them (it does happen) and then applied `font-size: 1em;` to `h3` tag inside they would display at different sizes. 😢 

`vh` and `vw` at the moment can be applied to type or layout, I don't think there has been a standard set either way. I've used them for both and had some nice results.

There is also the `fr` fractional unit introduced with CSS GRID to be aware of.

Agree not to use `s` for transitions and animations, I rarely see that now either.

🎉 for a @humanmade/frontend team

**Comment by mikeselander** [comment]
Ah, ❤️ the new team Ryan, thanks! I'm totally fine with blacklisting as opposed to whitelisting on the font size.

I'll prepare a PR to this affect shortly.

**Comment by rmccue** [comment]
@kirstyburgoine As an example: in the pattern library, we have:

```css
h1 {
    font-size: 1.555555556rem;
}
h2 {
    font-size: 1.333333333rem;
}
```

In H2, we use the pattern library, so we inherit these styles. However, I want to make all text in the profile component smaller, so I apply the following rule:

```css
.sidebar {
    font-size: 0.8em;
}
```

However, because the headings are specified with `rem`, they do not size down with the rest of the text. (`p` has no special rule, so it does size down.) This means I now need way more rules to override headings so that the relative sizing of paragraphs and headings remains the same.

```css
.sidebar h1 {
    font-size: 1.244444445rem;
    /* Equivalent: font-size: 1.555555556em; in 0.8em context */
}
.sidebar h2 {
    font-size: 1.066666666rem;
    /* Equivalent: font-size: 1.333333333em; in 0.8em context */
}
```

(We actually need a lot more, because there are `@media` queries, and all the margins are specified as `rem` as well.)

Essentially, when using `rem`, you completely lose the cascading effects of font-size. In a lot of cases that's intentional (as you said, it becomes easier to reason about), but it removes the flexibility to have less important visual elements on screen, emphasise in hero elements, etc.

(You could argue that the sidebar shouldn't be sized down, but in the context of H2, we need information density in the sidebar as it's very space-constrained.)

**Comment by kirstyburgoine** [comment]
@rmccue ok yeah, I can why see that could be an issue. I'm not a fan of applying `font-size` to layout elements as an approach to controlling type sizing though. Technically there's nothing wrong with it. I think it tends to create more problems than it solves, but it is a personal preference really. Anyway, always happy to talk about this stuff, but this ticket is probably no longer the right place to do that. :)
